### PR TITLE
Document Anbox WebRTC data proxy service

### PR DESCRIPTION
**Pull request by adglkh**

### CODE REVIEW

File: howto/stream/oob-data.md [hunk 1]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
    anbox cloud <--  the received data
    ```
 
-### Anbox data proxy
+### Anbox WebRTC data proxy
 
 To build up the communication bridge between an Android application and the Anbox WebRTC platform, you must implement a service or daemon that is responsible for:
```
Comment (ru-fu) [inline howto/stream/oob-data.md:111]:
```suggestion
To build up the communication bridge between an Android application and the Anbox WebRTC platform, Anbox Cloud provides a system daemon named `anbox-webrtc-data-proxy`. This daemon is responsible for:
```

File: howto/stream/oob-data.md [hunk 2]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
 
  - The socket connection
  - Polling events
  - Forwarding data between both ends
 
-A system service named `anbox-data-proxy`, which covers all the above functionalities, will be implemented in the 1.16.0 release of Anbox Cloud. It makes use of [`gbinder`](https://github.com/mer-hybris/libgbinder) to set up a communication bridge and provide reliable and high-efficiency data transmission between the Android system and the Anbox WebRTC platform.
+Meanwhile, a system daemon named `anbox-webrtc-data-proxy` running in the Anbox container registers an Android system service `org.anbox.webrtc.IDataProxyService` which allows Android developers to easily make use of binder IPC for data communication between an Android application and Anbox WebRTC platform through a file descriptor.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:117]:
```suggestion
The `anbox-webrtc-data-proxy` system daemon runs in the Anbox container and registers an Android system service named `org.anbox.webrtc.IDataProxyService`. This service allows Android developers to easily make use of binder interprocess communication (IPC) for data communication between an Android application and the Anbox WebRTC platform through a file descriptor.
```

File: howto/stream/oob-data.md [hunk 3]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+[note type="information" status="Note"]To interact with `org.anbox.webrtc.IDataProxyService` system service, the Android application must be [installed as a system app](https://discourse.ubuntu.com/t/how-to-install-an-apk-as-a-system-app/27086). [/note]
```
Comment (ru-fu) [inline howto/stream/oob-data.md:119]:
```suggestion
[note type="information" status="Note"]To interact with the `org.anbox.webrtc.IDataProxyService` system service, the Android application must be [installed as a system app](https://discourse.ubuntu.com/t/how-to-install-an-apk-as-a-system-app/27086). [/note]
```

File: howto/stream/oob-data.md [hunk 4]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+
+#### Access Data Proxy Service
+
+There are two ways to access the `org.anbox.webrtc.IDataProxyService` from an Android application.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:124]:
```suggestion
#### Access the data proxy dervice

There are two ways to access the `org.anbox.webrtc.IDataProxyService` from an Android application:
```

File: howto/stream/oob-data.md [hunk 5]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+- If you develop the application with Android studio, you can access the service by using the Android's reflection API.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:126]:
```suggestion
- If you develop the application with Android studio, you can access the service by using Android's reflection API.
```

File: howto/stream/oob-data.md [hunk 6]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+            e.printStackTrace();
+        } catch (InvocationTargetException e) {
+            e.printStackTrace();
+        } catch (ClassNotFoundException e) {
+            e.printStackTrace();
+        }
+
+        return service;
+    }
+```
```
Comment (ru-fu) [inline howto/stream/oob-data.md:146]: This code block must be indented to be part of the bullet list.
(If this causes issues with the spell check, you might need to remove the empty line in the code.)

File: howto/stream/oob-data.md [hunk 7]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+- Or if you ship the Android Application inside of AOSP source tree and [build](https://source.android.com/docs/setup/build/building) it form there, you could utilise the Android's hidden API to do so.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:148]:
```suggestion
- If you ship the Android application inside of the AOSP source tree and [build](https://source.android.com/docs/setup/build/building) it from there, you can use Android's hidden API to access the service.
```

File: howto/stream/oob-data.md [hunk 8]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+```
+    IBinder getDataProxyService() {
+        return ServiceManager.getService("org.anbox.webrtc.IDataProxyService");
+    }
+```
```
Comment (ru-fu) [inline howto/stream/oob-data.md:154]: Must be indented.

File: howto/stream/oob-data.md [hunk 9]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+#### Connect Data Channel
+
+To fetch the file descriptor referring to one data channel, send a request to the data proxy service through a binder transaction:
```
Comment (ru-fu) [inline howto/stream/oob-data.md:158]:
```suggestion
#### Connect the data channel

To fetch the file descriptor that refers to one data channel, send a request to the data proxy service through a binder transaction:
```

File: howto/stream/oob-data.md [hunk 10]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+            Log.e(TAG, "Failed to connect data channel '" +  channel + "': " + ex.getMessage());
+        } finally {
+            data.recycle();
+            reply.recycle();
+        }
+```
+
+#### Receive data from Anbox WebRTC platform
+
+Once the valid file descriptor is returned, launch an asynchronous task to read data from Anbox WebRTC platform:
```
Comment (ru-fu) [inline howto/stream/oob-data.md:186]:
```suggestion
#### Receive data from the Anbox WebRTC platform

Once the valid file descriptor is returned, launch an asynchronous task to read data from the Anbox WebRTC platform:
```

File: howto/stream/oob-data.md [hunk 11]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+        }
+
+        return null;
+    }
+}
+```
+
+#### Send data to Anbox WebRTC platform
+
+To send data to Anbox WebRTC platform through the file descriptor:
```
Comment (ru-fu) [inline howto/stream/oob-data.md:222]:
```suggestion
#### Send data to the Anbox WebRTC platform

To send data to the Anbox WebRTC platform through the file descriptor:
```

File: howto/stream/oob-data.md [hunk 12]
```diff
@@ -106,15 +106,132 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+    OutputStream ostream = new FileOutputStream(mFd.getFileDescriptor());
+    try {
+        ostream.write(data.getBytes(), 0, data.length());
+    } catch (IOException ex) {
+        Log.i(TAG, "Failed to write data: " + ex.getMessage());
+        ex.printStackTrace();
+    }
+```
+
+For a complete Android example, see the [out_of_band_v2](https://github.com/anbox-cloud/anbox-streaming-sdk/tree/master/examples/android/out_of_band_v2) project for more details.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:234]:
```suggestion
For a complete Android example, see the [out_of_band_v2](https://github.com/anbox-cloud/anbox-streaming-sdk/tree/master/examples/android/out_of_band_v2) project.
```

**Review by ru-fu — COMMENTED** [review]
Several nitpicks, but overall this looks good!

File: howto/stream/oob-data.md [hunk 13]
```diff
@@ -106,15 +106,131 @@ A trivial example to simulate the data transmission between an Anbox container a
...
 
-To build up the communication bridge between an Android application and the Anbox WebRTC platform, you must implement a service or daemon that is responsible for:
+To build up the communication bridge between an Android application and the Anbox WebRTC platform, Anbox Cloud provides a system daemon named `anbox-webrtc-data-proxy`. This daemon is responsible for:
 
- - The socket connection
- - Polling events
- - Forwarding data between both ends
+ - accept connection requests from an Android application
+ - connect to one specific data channel via Unix domain socket exposed by the Anbox WebRTC platform
+ - pass the connected socket as a file descriptor to the Android application
```
Comment (ru-fu) [inline howto/stream/oob-data.md:115]:
```suggestion
 - Accepting connection requests from an Android application
 - Connecting to one specific data channel via the Unix domain socket exposed by the Anbox WebRTC platform
 - Passing the connected socket as a file descriptor to the Android application
```

File: howto/stream/oob-data.md [hunk 14]
```diff
@@ -106,15 +106,131 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+OutputStream ostream = new FileOutputStream(mFd.getFileDescriptor());
+try {
+    ostream.write(data.getBytes(), 0, data.length());
+} catch (IOException ex) {
+    Log.i(TAG, "Failed to write data: " + ex.getMessage());
+    ex.printStackTrace();
+}
+```
+
+For a complete Android example, see the [out_of_band_v2](https://github.com/anbox-cloud/anbox-streaming-sdk/tree/master/examples/android/out_of_band_v2) project.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:233]: The link doesn't exist yet - I guess the project will be merged before we publish?

**Review by ru-fu — APPROVED** [review]

File: howto/stream/oob-data.md [hunk 15]
```diff
@@ -106,15 +106,169 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+        <receiver
+            android:name=".DataChannelEventReceiver"
+            android:enabled="true"
+            android:exported="true">
+            <intent-filter>
+                <action android:name="com.canonical.anbox.BROADCAST_DATA_CHANNELS_STATUS"/>
+            </intent-filter>
+        </receiver>
+
+    ```
```
Comment (ru-fu) [inline howto/stream/oob-data.md:135]: You need to use either indentation or ` ``` `, not both. For sections like this, we usually use the quotes, thus:

````suggestion
```
<receiver
    android:name=".DataChannelEventReceiver"
    android:enabled="true"
    android:exported="true">
    <intent-filter>
         <action android:name="com.canonical.anbox.BROADCAST_DATA_CHANNELS_STATUS"/>
    </intent-filter>
</receiver>
```
````

File: howto/stream/oob-data.md [hunk 16]
```diff
@@ -106,15 +106,169 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+Then a broadcast which carries the following parameters is sent out to the Android application on the change in the availability of data channels.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:137]:
```suggestion
Whenever the availability of data channels changes, a broadcast is sent out to the Android application. The broadcast contains the following parameters:
```

File: howto/stream/oob-data.md [hunk 17]
```diff
@@ -106,15 +106,169 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+| Parameters               | Type                                            | Description                                                                                                                                                                |
+| ------------------------ | ----------------------------------------------- | -----------------------------------------------                                                                                                                            |
+| `event`                  | String                                          | Can be `"created"` or `"destroyed"` respectively denotes the data channels are created and open for Android applications to use, or the data channels are closed and destroyed. |
```
Comment (ru-fu) [inline howto/stream/oob-data.md:141]:
```suggestion
| `event`                  | string                                          | Can be `created` (which means the data channels are created and open for Android applications to use) or `destroyed` (which means that the data channels are closed and destroyed) |
```

Comment (ru-fu) [reply howto/stream/oob-data.md:141]: I assume that the strings don't contain the double quotes? So it's, for example, `created` and not `"created"`?

File: howto/stream/oob-data.md [hunk 18]
```diff
@@ -106,15 +106,169 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+| `data-channel-names`     | String array                                    | Comma-separated list of data channel names that identify the changed data channels.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:142]:
```suggestion
| `data-channel-names`     | string array                                    | Comma-separated list of data channel names that identify the changed data channels
```

File: howto/stream/oob-data.md [hunk 19]
```diff
@@ -106,15 +106,169 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+The Android application implements a subclass of the [`BroadcastReceiver`](https://developer.android.com/guide/components/broadcasts#effects-process-state), which responds to the above events that are sent by the Android system.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:144]: We are giving instructions here for how users should implement their Android application, correct?
Then it should be "should implement", not "implements".

File: howto/stream/oob-data.md [hunk 20]
```diff
@@ -106,15 +106,169 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+
+    ```
```
Comment (ru-fu) [inline howto/stream/oob-data.md:146]: Same comment about the indentation.

**Review by ru-fu — CHANGES_REQUESTED** [review]

**Comment by adglkh** [comment]
@ru-fu Thanks for your comments, could you take another look?

**Review by ru-fu — APPROVED** [review]
Looks good, thanks! :)

File: howto/stream/oob-data.md [hunk 21]
```diff
@@ -10,7 +10,7 @@ Anbox Cloud provides two versions of this OOB data exchange:
   This version supports only half-duplex data transmission. It allows sending data from an Android application to a WebRTC client through the Anbox WebRTC platform, but it is not possible to receive data from the WebRTC client to an Android application.
 
 [note type="caution" status="Warning"]
-[Version 1](#oob-v1) of the out-of-band data exchange between an Android application and a WebRTC client is deprecated. Support for it will be discontinued in a future release. Therefore, you should migrate your integration of version 1 of the OOB data exchange to [version 2](#oob-v2) for full-duplex data transmission and better performance.
+The support to [version 1](#oob-v1) of the out-of-band data exchange between an Android application and a WebRTC client has been removed in Anbox Cloud 1.16 release. Therefore, you should migrate your integration of version 1 of the OOB data exchange to [version 2](#oob-v2) for full-duplex data transmission and better performance.
```
Comment (ru-fu) [inline howto/stream/oob-data.md:13]:
```suggestion
The support for [version 1](#oob-v1) of the out-of-band data exchange between an Android application and a WebRTC client has been removed in the Anbox Cloud 1.16 release. Therefore, you should migrate your integration of version 1 of the OOB data exchange to [version 2](#oob-v2) for full-duplex data transmission and better performance.
```

File: howto/stream/oob-data.md [hunk 22]
```diff
@@ -106,15 +106,168 @@ A trivial example to simulate the data transmission between an Anbox container a
...
+} catch (IOException ex) {
+    Log.i(TAG, "Failed to write data: " + ex.getMessage());
+    ex.printStackTrace();
+}
+```
+
+For a complete Android example, see the [out_of_band_v2](https://github.com/anbox-cloud/anbox-streaming-sdk/tree/master/examples/android/out_of_band_v2) project.
 
 <a name="oob-v1"></a>
 ## Version 1
```
Comment (ru-fu) [inline howto/stream/oob-data.md:273]: Please add a similar note to the one we have on top here in this section:

```suggestion
## Version 1

[note type="caution" status="Warning"]
The support for version 1 of the out-of-band data exchange between an Android application and a WebRTC client has been removed in the Anbox Cloud 1.16 release.
[/note]
```

**Review by ru-fu — CHANGES_REQUESTED** [review]

File: howto/stream/oob-data.md [hunk 23]
```diff
@@ -106,19 +106,175 @@ A trivial example to simulate the data transmission between an Anbox container a
...
 
+[note type="caution" status="Warning"]
+The support for version 1 of the out-of-band data exchange between an Android application and a WebRTC client has been removed in the Anbox Cloud 1.16 release.
+[/note]
```
Comment (ru-fu) [inline howto/stream/oob-data.md:277]: We'll need an empty line after the note (otherwise it won't show up correctly).

Comment (adglkh) [reply howto/stream/oob-data.md:277]: Thanks
Fixed.

**Review by ru-fu — APPROVED** [review]

**Pull request merged**